### PR TITLE
honor-first-shutdown-request

### PR DESCRIPTION
**Pull request by ejb42**
@keszybz 
Here is the new pull request I spoke of in the previous one I just closed.
I went ahead with the use of log_info instead of using log_unit_info since using manager_state() instead of manager_get_unit() precludes the use of log_unit_info. If you wish that instead I can change it but I think the use of manager_state is cleaner.
I also added a test in the test directory which incorporates the use of the test service file and script.

### CODE REVIEW

**Comment by ejb42** [comment]
@keszybz Is there anything I need to do for this pull request? Is there something wrong?

File: src/core/emergency-action.c [hunk 1]
```diff
@@ -32,6 +44,15 @@ void emergency_action(
         assert(action >= 0);
         assert(action < _EMERGENCY_ACTION_MAX);
 
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        if ((manager_state(m) == MANAGER_STOPPING) &&
+            ((action == EMERGENCY_ACTION_REBOOT) ||
+             (action == EMERGENCY_ACTION_POWEROFF))) {
+                log_info("EmergencyAction: Shutdown is already active Skipping %s request",
```
Comment (poettering) [inline src/core/emergency-action.c]: hmm, so manager_state() does a lot of things internally, most of which aren't interesting here. I think I would make the test differently:

```c
Unit *u;
u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
if (u && unit_active_or_pending(u) &&
        IN_SET(action, EMERGENCY_ACTION_REBOOT, EMERGENCY_ACTION_POWEROFF, EMERGENCY_ACTION_EXIT)) {
…
```

Also, I think this should be log_notice(), not log_info()

File: src/test/test-honor-first-shutdown.service [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Honor First Shutdown feature
+After=multi-user.target
+
+[Service]
+ExecStart=/test-honor-first-shutdown.sh
+ExecStop=pkill -9 test-honor-first-shutdown.sh
```
Comment (poettering) [inline src/test/test-honor-first-shutdown.service]:
```
ExecStop=kill -KILL $MAINPID
```

should just work, and not pull in pkill

File: src/core/emergency-action.c (hunk 1 above)
Comment (poettering) [reply src/core/emergency-action.c]: ANd yeah, EMERGENCY_ACTION_EXIT should be covered too here, it's just another way to shutdown (one you would only use in a container, but that's a detail)

File: src/test/test-honor-first-shutdown.service (hunk 2 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.service]: This pkill was the method I was using to cause termination of the loop inside the script and to cause the FailureAction=reboot emergency action during the shutdown. Otherwise the shutdown would just SIGTERM the script and no failure would be seen correct? I was thinking pkill might be a handy tool to have for the test environment. In any case, I will work on a different way so I won't have to pull in pkill.

**Comment by ejb42** [comment]
> 
> 
> Hmm, so you are only doing this for shutdowns initiated via the emergency action logic. That is intended, yeah? I think I am fine with this, I am just wondering if you are aware that this does not affect "systemctl poweroff" and so on, as those are regular, human requested shutdowns and not emergency ones...

My idea was that a shutdown has already been triggered via human or other way and the fix was to prevent another shutdown, via the emergency path, from interrupting that. I was thinking it would be rather difficult to manually initiate a second shutdown if one is in progress. I suppose someone could have some service that could have an ExecStop that tries to do a systemctl reboot(etc). Would that be something you would want to guard against or should we leave that to user error?

Thank you for your comments, I will work on your change suggestions.

File: src/test/test-honor-first-shutdown.service (hunk 2 above)
Comment (poettering) [reply src/test/test-honor-first-shutdown.service]: the line above is equivalent to the pkill. It uses the `$MAINPID` env var systemd passes to ExecStop= calls anyway, so the pkill is not necessary. The effect of the line I proposed and yours i the same: signal 9 (SIGKILL) is sent to the process invoked via ExecStart=

**Comment by poettering** [comment]
I think your approach is fine. If people place a "systemctl reboot" in ExecStop= they get to live with the effect...

File: src/test/test-honor-first-shutdown.service (hunk 2 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.service]: Ah, got it. Will do. Thanks

**Comment by ejb42** [comment]
@poettering, I believe I have addressed all of your comments, please review when you get a chance.

File: src/core/emergency-action.c [hunk 3]
```diff
@@ -12,6 +12,18 @@
...
+static const char* const emergency_action_table[_EMERGENCY_ACTION_MAX] = {
+        [EMERGENCY_ACTION_NONE] = "none",
+        [EMERGENCY_ACTION_REBOOT] = "reboot",
+        [EMERGENCY_ACTION_REBOOT_FORCE] = "reboot-force",
+        [EMERGENCY_ACTION_REBOOT_IMMEDIATE] = "reboot-immediate",
+        [EMERGENCY_ACTION_POWEROFF] = "poweroff",
+        [EMERGENCY_ACTION_POWEROFF_FORCE] = "poweroff-force",
+        [EMERGENCY_ACTION_POWEROFF_IMMEDIATE] = "poweroff-immediate",
+        [EMERGENCY_ACTION_EXIT] = "exit",
+        [EMERGENCY_ACTION_EXIT_FORCE] = "exit-force",
```
Comment (keszybz) [inline src/core/emergency-action.c]: Please align the rhs values to the same column. It's much easier to scan by eye then.

File: src/core/emergency-action.c [hunk 4]
```diff
@@ -28,10 +40,22 @@ void emergency_action(
...
         assert(action >= 0);
         assert(action < _EMERGENCY_ACTION_MAX);
 
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if (u && unit_active_or_pending(u) &&
+            IN_SET(action, EMERGENCY_ACTION_REBOOT,
+                   EMERGENCY_ACTION_POWEROFF, EMERGENCY_ACTION_EXIT)) {
+                log_notice("EmergencyAction: Shutdown is already active Skipping %s request",
+                         emergency_action_table[action]);
```
Comment (keszybz) [inline src/core/emergency-action.c]: Indentation, and the sentences run together. Maybe "Shutdown is already active. Skipping emergency action request %s."

File: src/test/test-honor-first-shutdown.service [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Honor First Shutdown feature
+After=multi-user.target
+
+[Service]
+ExecStart=/test-honor-first-shutdown.sh
+ExecStop=/bin/sh -x -c 'kill -SIGKILL $MAINPID'
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.service]: The shell is not nedeed here. Just `ExecStop=kill -SIGKILL $MAINPID` .

File: src/test/test-honor-first-shutdown.sh [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+echo "Honor first shutdown test script"
+while true; do
+    sleep 3;
+done
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.sh]: Maybe `sleep infinity` instead of the loop?

File: test/TEST-48-HONORFIRSTSHUTDOWN/test.sh [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    mask_supporting_services
+
+    # setup honor first shutdown service
+    cp ../../src/test/test-honor-first-shutdown.service $initdir/etc/systemd/system/
+    cp ../../src/test/test-honor-first-shutdown.sh $initdir/
+    chmod 755 $initdir/test-honor-first-shutdown.sh
+
+    # setup the testsuite service
+    cp ../testsuite-48.units/testsuite-48.service $initdir/etc/systemd/system
+    cp ../testsuite-48.units/testsuite-48.sh $initdir/
```
Comment (keszybz) [inline test/TEST-48-HONORFIRSTSHUTDOWN/test.sh]: Please please just do what the other tests do. There is no need to copy files.

File: test/test-functions [hunk 8]
```diff
@@ -389,9 +389,9 @@ run_nspawn() {
         dwarn "nspawn doesn't support SYSTEMD_NSPAWN_UNIFIED_HIERARCHY=hybrid, skipping"
         exit
     elif [[ "$UNIFIED_CGROUP_HIERARCHY" = "yes" || "$UNIFIED_CGROUP_HIERARCHY" = "no" ]]; then
-        _nspawn_pre=("${nspawn_pre[@]}" env SYSTEMD_NSPAWN_UNIFIED_HIERARCHY=$UNIFIED_CGROUP_HIERARCHY)
+        _nspawn_pre=("${_nspawn_pre[@]}" env SYSTEMD_NSPAWN_UNIFIED_HIERARCHY=$UNIFIED_CGROUP_HIERARCHY)
     elif [[ "$UNIFIED_CGROUP_HIERARCHY" = "default" ]]; then
-        _nspawn_pre=("${nspawn_pre[@]}" env --unset=UNIFIED_CGROUP_HIERARCHY --unset=SYSTEMD_NSPAWN_UNIFIED_HIERARCHY)
+        _nspawn_pre=("${_nspawn_pre[@]}" env --unset=UNIFIED_CGROUP_HIERARCHY --unset=SYSTEMD_NSPAWN_UNIFIED_HIERARCHY)
```
Comment (keszybz) [inline test/test-functions]: This bug should be fixed in a separate commit.

File: test/testsuite-48.units/testsuite-48.service [hunk 9]
```diff
@@ -0,0 +1,6 @@
+[Unit]
+Description=Testsuite service
+
+[Service]
+ExecStart=/testsuite-48.sh
```
Comment (keszybz) [inline test/testsuite-48.units/testsuite-48.service]: No files in root. Just do what the other tests do.

**Comment by poettering** [comment]
please generate "perfect" PRs, i.e. where each individual commit is a logical step that makes sense, and not a historical one. We want perfect bisectability, and that means that commits in a PR shouldn't fix up commits in the same PR.

**Comment by poettering** [comment]
also, plese provide a useful commit msg in the first place

File: src/core/emergency-action.c [hunk 10]
```diff
@@ -28,10 +40,23 @@ void emergency_action(
...
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if (u && unit_active_or_pending(u) && IN_SET(action,
+                                                     EMERGENCY_ACTION_REBOOT,
+                                                     EMERGENCY_ACTION_POWEROFF,
+                                                     EMERGENCY_ACTION_EXIT)) {
+                log_notice("Shutdown is already active. Skipping emergency action request %s.",
+                           emergency_action_table[action]);
+                return;
+        }
```
Comment (poettering) [inline src/core/emergency-action.c:57]: hmm, can you reverse the check? manager_get_unit() is more expensive than just comparing the action. hence maybe do it like this:

```c
if (IN_SET(action, EMERGENCY_ACTION_REBOOT, EMERGENCY_ACTTION_POWEROFF, EMERGENCY_ACTION_EXIT)) {
        Unit *u;
        u = manager_get_unit(m, …);
        if (u && unit_active_or_pending(u)) {
                …
```

File: src/test/test-honor-first-shutdown.service (hunk 5 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.service]: @keszybz I have tried to use this ExecStop without the shell but simply cannot make it work. I found that testsuite-09 uses the same shell to implement a kill in it's ExecStop. When I remove the shell and run the test I get the following error when running the image.

TEST-48-HONORFIRSTSHUTDOWN RUN: testing honor first shutdown
unit_file_build_name_map: normal unit file: /usr/lib/systemd/tests/testdata/units/test-honor-first-shutdown.service
test-honor-first-shutdown.service: Trying to enqueue job test-honor-first-shutdown.service/start/replace
Sent message type=error sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=org.freedesktop.systemd1.BadUnitSetting error-message=Unit test-honor-first-shutdown.service has a bad unit file setting.
Failed to process message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartUnit cookie=1 reply_cookie=0 signature=ss error-name=n/a error-message=n/a: Unit test-honor-first-shutdown.service has a bad unit file setting.

I changed it to be more like what is in testsuite-09 as follows:
ExecStop=sh -c 'kill -SIGKILL $MAINPID'

If using the shell operation is not acceptable please help me understand how to fix it.

**Comment by ejb42** [comment]
Additional note with this push: The tests have been failing due to a timeout issue. It may be due to my fork being so far behind the source. I will look into any failures that may occur with the current tests.

File: src/core/emergency-action.c (hunk 10 above)
Comment (ejb42) [reply src/core/emergency-action.c:57]: This is done in latest PR

**Comment by ejb42** [comment]
@poettering @keszybz The last commit was done to fix a numbering conflict of the test files with the main repo.

**Comment by keszybz** [comment]
LGTM. I think the test could be still simplified a bit, but this isn't terribly important. Let's merge.
